ariadne_graphql_modules/typing.py

```diff
@@ -0,0 +1,157 @@
+)
+
+from graphql import (
+    ListTypeNode,
+    NamedTypeNode,
+    NameNode,
+    NonNullTypeNode,
+    TypeNode,
+)
+
+from ariadne_graphql_modules.base import GraphQLMetadata, GraphQLType
+from ariadne_graphql_modules.deferredtype import DeferredTypeData
+from ariadne_graphql_modules.idtype import GraphQLID
+
+if sys.version_info >= (3, 10):
+    from types import UnionType
+else:
+    UnionType = Union
+
+
+def get_type_node(  # noqa: C901
+    metadata: GraphQLMetadata,
+    type_hint: Any,
+    parent_type: Optional[type[GraphQLType]] = None,
+) -> TypeNode:
+    if is_nullable(type_hint):
+        nullable = True
+        type_hint = unwrap_type(type_hint)
+    else:
+        nullable = False
+
+    type_node: Optional[TypeNode] = None
+    if is_list(type_hint):
+        list_item_type_hint = unwrap_type(type_hint)
+        type_node = ListTypeNode(
+            type=get_type_node(metadata, list_item_type_hint, parent_type=parent_type)
+        )
+    elif type_hint is str:
+        type_node = NamedTypeNode(name=NameNode(value="String"))
+    elif type_hint is int:
+        type_node = NamedTypeNode(name=NameNode(value="Int"))
+    elif type_hint is float:
+        type_node = NamedTypeNode(name=NameNode(value="Float"))
+    elif type_hint is bool:
+        type_node = NamedTypeNode(name=NameNode(value="Boolean"))
+    elif get_origin(type_hint) is Annotated:
+        forward_ref, type_meta = get_args(type_hint)
+        if not type_meta or not isinstance(type_meta, DeferredTypeData):
+            raise ValueError(
+                f"Can't create a GraphQL return type for '{type_hint}'. "
+                "Second argument of 'Annotated' is expected to be a return "
+                "value from the 'deferred()' utility."
+            )
+
+        deferred_type = get_deferred_type(type_hint, forward_ref, type_meta)
+        type_node = NamedTypeNode(
+            name=NameNode(value=metadata.get_graphql_name(deferred_type)),
+        )
+    elif isinstance(type_hint, ForwardRef):
+        type_name = type_hint.__forward_arg__
+        if not parent_type or parent_type.__name__ != type_name:
+            raise ValueError(
+                "Can't create a GraphQL return type"
+                f"for forward reference '{type_name}'."
+            )
+
+        type_node = NamedTypeNode(
+            name=NameNode(value=metadata.get_graphql_name(parent_type)),
+        )
+    elif isinstance(type_hint, str):
+        type_node = NamedTypeNode(
+            name=NameNode(value=type_hint),
+        )
+    elif isclass(type_hint):
+        if issubclass(type_hint, GraphQLID):
+            type_node = NamedTypeNode(name=NameNode(value="ID"))
+        elif issubclass(type_hint, (GraphQLType, Enum)):
+            type_node = NamedTypeNode(
+                name=NameNode(value=metadata.get_graphql_name(type_hint)),
+            )
+
+    if not type_node:
+        raise ValueError(f"Can't create a GraphQL return type for '{type_hint}'.")
```

Comment:
In testing I've found that those errors can be very unspecific. `get_type_node` could accept additional kwarg like `location: str` that would go to this error, so we could make it "Can't create a GraphQL return type for type_hint in `SomethingSomethingType`".